Introduce an agent in libmcount

clementguidi opened the pull request:
Hi,

This is a rework of PR #1269 from DORSAL. I moved the code to a new repository and made modifications according to old discussions. Sorry for the inconvenience.

This PR is the first step to interact with uftrace targets at runtime. It separates the CLI from the libmcount runtime, by spawning a daemon in libmcount, that is controlled by an separate and independent uftrace instance.

I introduce two new associated parameters:
- `--daemon`/`-m` to spawn the daemon in the target (off by default)
- `--pid`/`-p`to set uftrace to "client mode" and forward its parameters to a running target

Note that the daemon code is compiled conditionally in libmcount. It can be disabled using the `--without-daemon` flag in the configure script.

Follow-up steps:
- implement parameter tweaking at runtime (e.g. filters, depth) - PR to come
- implement dynamic patching and unpatching (from #1274)

Please let me know what you think about it!

honggyukim commented:
Hi @clementguidi, thanks for coming back. I understand you want to make runtime dynamic patching. Is it correct?

One of our long term plan is to support attaching to a running process. It can be done together with runtime dynamic patching. I've summarized the approach at https://github.com/namhyung/uftrace/issues/1414#issuecomment-1081313032. You could have a look at it.

I've also uploaded making `libmcount.so` linked to a target binary at #1526. I think we can ask users to build the target program with explicitly linking `libmcount.so` and control whether it keeps socket open or not with an environmental variable such as `UFTRACE_TRACING_ON`.

This work is not same as your approach, but I think we can take advantage of both approaches. Thanks.

namhyung commented:
@clementguidi thanks for working on this!

Actually I was skeptical about the daemon work at first, but now I think it might be good both for changing tracing configurations at run time and for attaching to a task as an interim step.  A few thoughts:

* It's not a daemon actually, can we have a better name?
* most of socket messaging code should be in a separate file under utils directory
* please split the -m and -p option handling into separate commits
* I don't like the idea of optional argument, please make it required
* I don't think it needs a config option to disable it as it's already optional

clementguidi commented:
I made changes according to your comments first @namhyung.

As a new name for the daemon I suggest using "agent", what do you think? Other choices might be "controller" or "session", but the latter can be ambiguous in uftrace.

namhyung commented:
'agent' looks good.  Will review the patch soon.

namhyung commented on utils/socket.h:
```diff
@@ -0,0 +1,14 @@
+
+int create_socket(struct sockaddr_un *addr, pid_t pid);
+int listen_socket(int sfd, struct sockaddr_un *addr);
+int send_options(struct uftrace_opts *opts);
```
For these external functions, I prefer having a common prefix.  How about `socket_create`, `socket_listen`, `socket_send_options`?  It'd be nice if we could have some comments for them.

namhyung commented on utils/socket.h:
```diff
@@ -0,0 +1,14 @@
+
+#include <sys/socket.h>
+
+#include "uftrace.h"
```
Please don't include this unnecesarily.  If it's only for `struct uftrace_opts`, we can have a forward declaration.

namhyung commented on utils/socket.c:
```diff
@@ -0,0 +1,65 @@
+	char *channel = NULL;
+	struct sockaddr_un addr;
+
+	sfd = create_socket(&addr, opts->pid);
```
Please check the return value.

namhyung commented on libmcount/mcount.c:
```diff
@@ -1716,6 +1723,133 @@ static void mcount_script_init(enum uftrace_pattern_type patt_type)
+}
+
+/* Agent routine, applying instructions from the CLI. */
+void *apply_client_commands(void *arg)
```
It'd be nice we could have `agent_` prefix.  How about `agent_apply_commands()`?

namhyung reviewed (commented):
Could you please add a test case?  You could add a test command to check if it's communicating properly.

clementguidi commented:
Thank you very much for all your comments, I made the relevant changes.

As for the test case, I'm not sure how to do it. I've been trying to add a unit test in `uftrace.c` but I can't configure `libmcount` so it starts the agent when running that test.

namhyung commented on uftrace.c:
```diff
@@ -703,6 +705,11 @@ static int parse_option(struct uftrace_opts *opts, int key, char *arg)
 
+	case 'p':
+		opts->pid = strtol(arg, NULL, 0);
+		opts->exename = "";
```
Could you explain why this is needed?

namhyung replied:
Maybe we can add this functionality to `live` command and skip recording/replaying.

namhyung replied:
Ok, we don't need to use `record` for this.  I think `live` is a better fit since it'll only talk to a running process.  It can handle this similar to `--list-event`.  Wdyt?

clementguidi replied:
I agree, and it also means it can be invoked without explicitly using the `live` command.

As for the syntax, should I keep the `-p` option to trigger client behavior? I was thinking maybe use the target program as the PID value when it is numeric. For example
```
$ uftrace --some-option target # executes and traces the "target" binary
$ uftrace --some-option 1234   # client mode, forwards options to agent in process with PID 1234
```
Would that be relevant?

namhyung replied:
I think it's better to use `-p` explicitly to trigger the client mode.

namhyung reviewed (commented):
You can add a runtime test as similar as your use case.  You might want to add `UFTRACE_DOPT_TEST` to verify agent behavior in some way.

namhyung reviewed (commented):
The first commit contains changes for later commits.  Please split the first and squash them to the second and third so that each option comes together with the actual implementation.

clementguidi commented:
I added a test case to check if the agent is responsive. I'm using a structure similar to the `recv` tests, where you spawn a background uftrace instance and terminate it at the end of the test.

I split the commits as you requested.

namhyung reviewed (commented):
The test 272 was taken by mermaid recently, please update.  Also we might want to add more tests for agent later.  Please rename it to something like '273_agent_basic'.  Otherwise, looks good! :)

clementguidi commented:
Great! I'm fixing this and rebasing.

namhyung commented on libmcount/mcount.c:
```diff
@@ -1716,6 +1724,145 @@ static void mcount_script_init(enum uftrace_pattern_type patt_type)
 }
 
+/* unlink socket file if it exists */
+static void unlink_socket_maybe(struct sockaddr_un *addr)
```
`socket_unlink` and move it to socket.c?

namhyung commented on libmcount/mcount.c:
```diff
@@ -1716,6 +1724,145 @@ static void mcount_script_init(enum uftrace_pattern_type patt_type)
+
+	if (mkdir(MCOUNT_AGENT_SOCKET_DIR, 0775) == -1) {
+		if (errno != EEXIST) {
+			pr_dbg("error creating run directory %s\n", MCOUNT_AGENT_SOCKET_DIR);
```
Isn't it a critical error?  I wonder if it's a part of `socket_create`.

clementguidi replied:
It is a critical error, so I'm making it return.

However `socket_create()` doesn't bind to the socket, so it doesn't interact with the filesystem.
`socket_listen` does (I gather `bind()` and `listen()` calls there), but I feel like verifying that the directory exists prior to attempting socket creation/binding.

namhyung replied:
I'm ok with checking EEXIST.  Other error code means it has a more serious error so it should fail.

namhyung commented on libmcount/mcount.c:
```diff
@@ -1716,6 +1724,145 @@ static void mcount_script_init(enum uftrace_pattern_type patt_type)
+	if (sfd == -1)
+		return sfd;
+
+	unlink_socket_maybe(addr);
```
Why do you want to remove it?

clementguidi replied:
Socket binding fails when the file already exists. The only case I see it happening is if a PID is used twice by a uftrace target, and isn't cleared by the first one.

namhyung replied:
Ok, then I think we should warn instead if the file exists.

namhyung commented on libmcount/mcount.c:
```diff
@@ -1716,6 +1724,145 @@ static void mcount_script_init(enum uftrace_pattern_type patt_type)
+	if (agent_run)
+		agent_run = false;
+	else
+		return;
```
I prefer having return condition separately like

```c
if (!agent_run)
    return;

agent_run = false;
```

namhyung commented on libmcount/mcount.c:
```diff
@@ -1716,6 +1724,145 @@ static void mcount_script_init(enum uftrace_pattern_type patt_type)
+		break;
+	default:
+		pr_dbg("agent left in unknown state\n");
+	}
```
Why do you handle EINVAL specially?  I think we can just add `pr_dbg()` for the any error return.

clementguidi replied:
It was in case the agent thread terminated by itself, but this shouldn't happen.

namhyung commented on utils/socket.c:
```diff
@@ -0,0 +1,67 @@
+		pr_warn("cannot connect to socket '%s': %s\n", addr->sun_path, strerror(errno));
+		return -1;
+	}
+	else
```
You can remove the else.

namhyung commented on cmds/live.c:
```diff
@@ -95,6 +97,39 @@ static void setup_child_environ(struct uftrace_opts *opts)
+		if (read(sfd, &ack, sizeof(enum uftrace_dopt)) == -1)
+			ret = -1;
+		else
+			ret = -!(ack == UFTRACE_DOPT_CLOSE);
```
Please change it to

```c
if (read(...) < 0 ||
    ack != UFTRACE_DOPT_CLOSE)
        ret = -1;
```

namhyung commented on tests/t273_agent_basic.py:
```diff
@@ -0,0 +1,37 @@
+        self.record_p.communicate(b"^D") # target waits for a char to end
+        self.record_p.wait()
+        if result == TestBase.TEST_DIFF_RESULT: # the expected result is empty
+            result = TestBase.TEST_SUCCESS
```
This is wierd... What's the actual output?

clementguidi replied:
I only want to test the exit value of `uftrace live --pid`, as it produces no trace output. That's the only way I could think of using the test setup: starting a target with an agent in the background, and check if the client (`uftrace live --pid`) can communicate with the agent.
The test doesn't expect an output from the `live` command, so `TEST_DIFF_RESULT` is returned. But I assume the test is successful as long as the exit code from `live` is 0.

namhyung replied:
I guess it can get the usual output with the agent and the expected output is something like below.  Then, why not just set the output and check the return code?

```
# DURATION     TID     FUNCTION
            [266409] | main() {
   2.530 us [266409] |   getchar();
   3.037 us [266409] | } /* main */
```

clementguidi replied:
Right, I wasn't sure how to use the test structure to start multiple processes. Now I check the client return code in the `prerun()` and use the `replay` command in the `run()` so I get a trace output.

namhyung reviewed (commented):
Nit: small style changes.

namhyung commented on tests/t273_agent_basic.py:
```diff
@@ -0,0 +1,37 @@
+
+from runtest import TestBase
+
+EXEARG = 't-agent'
```
I don't think we need this because it'd be set by default.

namhyung commented on tests/t273_agent_basic.py:
```diff
@@ -0,0 +1,43 @@
+        sleep(.05)              # time for the agent to start
+        self.subcmd = 'live'
+        self.option = '-p %d' % record_p.pid
+        self.exearg = ''
```
It'd nice if this command actually doing something to change the output.  I think we can change this test after we add some new features/operations later.

clementguidi replied:
Sure, I'll add tests in other pull requests.

namhyung reviewed (approved):
LGTM

clementguidi commented:
Great, thank you for your guidance!
